test/e2e/tests/phishing-controller/phishing-detection.spec.js

```diff
@@ -297,4 +332,151 @@ describe('Phishing Detection', function () {
   });
+
+  describe('Phishing redirect protections', function () {
+    /**
+     * Status codes 305 (via Location header) and 306 (Set-Proxy) header do not
+     * automatically forward user agents to the new location, so we don't test
+     * these status codes. Status code 304 (Not Modified) is not a redirect
+     * status code.
+     *
+     * 201 _does_ require a Location header, but it also doesn't get
+     * automatically redirected by the user-agent, so we don't test it either.
+     *
+     * 200 is only included in order to test the `Refresh` header.
+     *
+     * @type {readonly [200 | 301 | 302 | 303 | 307 | 308]}
+     */
+    const redirectableStatusCodes = [200, 301, 302, 303, 307, 308];
+
+    const destination = 'https://metamask.github.io/test-dapp/';
+    const blocked = '127.0.0.1';
+
+    /**
+     * @type {Server | undefined}
+     */
+    let server;
+
+    /**
+     * @type {Driver | undefined}
+     */
+    let driver;
+
+    /**
+     * @type {Promise<void> | undefined}
+     */
+    let fixturePromise;
+
+    /**
+     *  Handle requests by setting the given header values and status code.
+     *
+     * @param {string} name - The name of the header to set.
+     * @param {string} value - The value of the header to set.
+     * @param {typeof redirectableStatusCodes[number]} code - The status code
+     * @returns {void}
+     */
+    function handleRequests(name, value, code) {
+      server.once('request', async function (_request, response) {
+        response.setHeader(name, value).writeHead(code).end(`
+        <!doctype html>
+        <html>
+          <head>
+            <meta http-equiv="Refresh" content="0;url="${destination}"/>
+            <title>Phishing test</title>
+
+            <script>
+              // this script should not run.
+              // it is meant to test for regressions in our redirect
+              // protection due to changes in either MetaMask or browsers.
+              document.location.href = "${destination}";
+              alert("trying to prevent phishing protection");
+              while(true){}
+            </script>
+          </head>
+          <body>
+            <h1>Redirecting...</h1>
+          </body>
+        </html>
+        `);
+      });
+    }
+
+    before('Set up redirect server', async function () {
+      const { promise, resolve, reject } = createDeferredPromise();
+      server = createServer();
+      server.listen(0, blocked, resolve);
+      server.on('error', reject);
+      await promise;
+    });
+    after('Shut down redirect server', async function () {
+      if (server) {
+        const { promise, resolve } = createDeferredPromise();
+        server.close(() => resolve());
+        // We need to close all connections to stop the server quickly
+        // Otherwise it takes a few seconds for it to close
+        server.closeAllConnections();
+        await promise;
+      }
+    });
+    const deferredTestSuite = createDeferredPromise();
```

Comment:
Haha, wow that is an interesting solution to that problem.

It does make sense to enable using a single `withFixtures` call for multiple tests. Have you tried moving the tests inside the `withFixtures` callback instead? That might be a simpler way to do this.

Reply:
Yes, but then we'd be back to doing things The Wrong Way™️.

Test set up should be inside `before` or `beforeAll` and tear down in `after` or `afterAll`. Doing it the Right Way®️ will give us access to more granular test metrics and proper setup/tear down will prevent cascading test failures due to issue with open handles (when the test run kills a the test process for whatever reason, it _always_ runs the after hooks).

Perhaps `withFixtures` should be refactored to be used in it's current form, as well as in a form that makes the patterns in this PR less chaotic; maybe a `startFixtures` function that returns something like `Promise<{ driver, <... etc...>, end: () => Promise<void> }>`  where `end` could be called by the `after`/`afterAll`. Thoughts (and if yes, should it be done in this PR, before this PR, or as a follow up PR)?

Reply:
Hmm, our general preference in other test suites has been to use "setup functions" over hooks for setup steps, so that they can be configured on a per-test basis while still reusing the setup code. For example, that's what `withFixtures` does.  Another example of this is the `withController` setup function we have in various controller unit test suites. Our unit test contributor docs suggest this pattern as well (see [here](https://github.com/MetaMask/contributor-docs/blob/main/docs/unit-testing.md#avoid-the-use-of-beforeeach), in the "Read more" section)

This would still include "proper setup/tear down" because we can tear down in a `finally` block in the setup function.

Placing multiple tests within a `withFixtures` callback does break test isolation, but so does what you're doing here. And that seems like a reasonable tradeoff to make in an e2e test context.

tl;dr, maybe putting multiple tests in a `withFixtures` callback wouldn't be "The Wrong Way™️"?

I don't know what "granular test metrics" means in this context though, maybe there is some benefit to using hooks that I am missing there.

Reply:
> we can tear down in a finally block in the setup function.

There are cases where a test can be forcibly killed by the test runner (mocha, jest) and `finally` blocks are _not_ run, but `after` will be.

Not doing test cleanup is not desirable (I know you know, but others might read this too, so I'm saying it anyway), as it can leave files behind or port-bound processes running, like ganache or other servers -- and the next test that attempts to spin up these servers will also fail. This makes understanding debug logs a bit more cumbersome.

I've thought of two cases I've seen happen in the extension code base.

When a test times out a `finally` won't run:

```
// name this file test.js then run `npx -y mocha test.js`
describe("test", () => {
  before(()=> {
    console.log("before");
  });

  after(()=> {
    console.log("after");
  });

  function runTest() {
    return new Promise(resolve => {
      setTimeout(resolve, 10000);
    })
  }

  it("works", async () => {
    try {
      await runTest()
    } finally {
      console.log("finally"); // this never runs
    }
  })
})
```

Another case where `finally` is not sufficient is when an unhandled exception/rejection bubbles up to the top process. Example:

```javascript
// name this file test.js then run `npx -y mocha test.js`
describe("test", () => {
  before(()=> {
    console.log("before");
  });

  after(()=> {
    console.log("after");
  });

  function runTest() {

    setTimeout(() => {
      // pretend this is thrown by some listener we've got installed somewhere
      throw new Error("error");
    }, 100);


    return new Promise(resolve => {
      setTimeout(resolve, 1000);
    })
  }

  it("works", async () => {
    try {
      await runTest()
    } finally {
      console.log("finally"); // this will not be logged
    }
  })
})
```

Regarding granular timings: we can measure how long the actual tests take, vs the test setup and teardown. We can then more easily optimize for them. Not like it's useful right now, as we've got way bigger fish to fry, but if we wanted to start implementing upper bounds for test duration these timings could prove useful.